### ISSUE TITLE
Automation: "Get Permission"

### ISSUE THREAD

**Issue by jugglinmike** [w3c/permissions#180]
Currently, the Automation section defines a WebDriver extension command for "Set Permission," without a corresponding "Get Permission". We deferred specifying such a command because our motivation at the time was enabling automated testing of [powerful features](https://w3c.github.io/permissions/#powerful-feature), so state retrieval was not critical.

There are other use cases for automating the Permissions API, though, so we should talk about those and determine if they warrant a new WebDriver command.

Based on my current understanding of the typical application developer's use case, a "Get Permission" command may not be necessary. Folks could query the permission status at the granularity of an arbitrary permission descriptor by combining [WebDriver's "Execute Async Script" command](https://www.w3.org/TR/webdriver/#execute-async-script) with [the Permissions API's `query` method](https://w3c.github.io/permissions/#dom-permissions-query).

Here's what it could look like in Python:

    descriptor = {'name': 'camera', 'deviceID': '1234'}

    status = session.execute_async_script('''
        var descriptor = arguments[0];
        var done = arguments[1];

        navigator.permissions.query(descriptor)
          .then(function() {
            return {state: result.state};
          }, function(error) {
            return {error: error.message};
          }).then(done);
        ''')

@agouaillard-cosmo has a much clearer understanding of the needs than I, so they can say more about whether this is sufficient. What do you think, Dr. Alex?

**Comment by youennf** [comment]
Overall, I agree that the get API is probably not high priority .
I just want to mention that the permission API might not be sufficient, depending on how it is actually implemented.

In Safari, the prompt is made once for the lifetime of the execution context.
Once it is answered, any subsequent getUserMedia call will either be granted or denied as per this initial decision.

As Safari is implemented today (modulo query API that is not supported), the following would probably happen:
WebDriver sets permission to false.
Query API will return false.
WebDriver sets permission to true.
Query API will return true.
getUserMedia is called and succeeds - storage of this decision for the context is done.
WebDriver set permission to false.
Query API will "probably" return true since another getUserMedia call for that context will succeed.

**Comment by youennf** [comment]
> Overall, I agree that the get API is probably not high priority .
> If the approach I suggested was sufficient for all use cases, then the "Get Permissions" command wouldn't just be low priority; it would be unnecessary. Can you say more about the use cases which motivate the command?

I am not sure there is a precise use case except convenience. Since the app can retrieve these values, it frees the app to store these values itself.

> This sounds acceptable as long as the "Set Permission" command is rejected in accordance with the following algorithm step:
> 
> If parameters.state is an inappropriate permission state for any implementation-defined reason, return a WebDriver error with WebDriver error code invalid argument.

Well, it could somehow work if the set permission is focusing on a specific context.
This might get messy when the set permission is about several contexts.

At least in WebKit implementation, this would also be somehow a layer violation.
WebDriver is a replacement to what Safari does and does not know about WebKit engine internal states.

**Comment by jugglinmike** [comment]
> This might get messy when the set permission is about several contexts.

Since this feedback is about the existing "Set Permission" command, we've moved the discussion to a new issue: https://github.com/w3c/permissions/issues/181

In order to resolve this issue, we still need a use case that motivates a "Get Permission" command. If no one can identify such a use case after some time, then we should be able to close this issue without introducing any new specification text.

**Comment by foolip** [comment]
@guidou @henbos @kereliuk, I believe that for the purposes of automating getUserMedia or WebRTC tests in Chromium, querying the permission state from WebDriver (or JS) will not be necessary, can you confirm?

**Comment by kereliuk** [comment]
I believe this is the case if we are running tests from Chromium (not WPT).

There is still some implementation to be done, but we can leave the protocol specific parts out and the rest can be built on from within Chromium. That implementation work would is not blocked on WebDriver implementations.

**Comment by alvestrand** [comment]
Can we write tests against this API that will succeed on both Chrome (with all-cameras permission) and on Firefox (with one-camera-at-a-time permission)?

**Comment by jugglinmike** [comment]
@alvestrand I'm not sure how to interpret your question. Is one of these re-phrasings accurate, or is it something else?

- Can we write tests for other specifications using a new "Get Permissions" command [...]
- Can we write tests for other specifications using the current "Set Permissions" command [...]

**Comment by jan-ivar** [comment]
> WebDriver sets permission to false.
> Query API will "probably" return true since another getUserMedia call for that context will succeed.

@youennf permission is not boolean, but rather: `"denied"`, `"granted"` or `"prompt"`.

But good point. `"denied"` is supposed to block, so it would seem odd to ignore it.

At the same time, [getUserMedia]() says to: *"[consider] all devices attached to a live MediaStreamTrack in the current browsing context to have permission status "granted"*, so it wins.

Firefox gets around this by always `stop()`ing affected tracks in response a user revoking permission, but that's browser-specific of course.

> I am not sure there is a precise use case except convenience.

In that case, I vote we close this.

**Comment by jan-ivar** [comment]
Note that for [specific reasons](https://github.com/mozilla/standards-positions/issues/19#issuecomment-365152565) Firefox does not support `query()` for `"camera"` or `"microphone"`. In short, `query` exposes details of non-standard permission models, which we fear can lead web developers to code towards the dominant permission model (Chrome's), causing headaches for others.

If we were to implement it—which we're still discussing—we'd likely need to lie and return `"granted"` instead of `"prompt"` on all but the very first call (even though there will be a prompt), to mimic Chrome and avoid web developers inadvertently priming Firefox users every time, a veritable slow-lane.
